### PR TITLE
CE-510: remove doubled pickup counters

### PR DESCRIPTION
**Pull request by serishema**
CE-510.

### CODE REVIEW

File: src/engine/world/world.h [hunk 1]
```diff
@@ -292,6 +292,15 @@ class World final
 
   void addPickupWidget(Sprite sprite, size_t count)
   {
+    auto isSameItemAndSameAction = [sprite](ui::PickupWidget item) -> bool {
+      return item.m_sprite.textureId == sprite.textureId && item.m_duration == 75_frame;
```
Comment (stohrendorf) [inline src/engine/world/world.h:296]: `75_frame` should be extracted into a constant.

File: src/ui/pickupwidget.h [hunk 2]
```diff
@@ -44,7 +44,6 @@ class PickupWidget
     return m_sprite.render1.x - m_sprite.render0.x;
   }
 
-private:
```
Comment (stohrendorf) [inline src/ui/pickupwidget.h:47]: Please don't make stuff public like that. Add getters instead.

File: src/engine/world/world.h [hunk 3]
```diff
@@ -292,6 +292,15 @@ class World final
...
+    };
+    if(auto it = std::find_if(m_pickupWidgets.begin(), m_pickupWidgets.end(), isSameItemAndSameAction); 
+       it != std::end(m_pickupWidgets))
```
Comment (stohrendorf) [inline src/engine/world/world.h:299]:
```suggestion
       it != m_pickupWidgets.end())
```
That's inconsistent to how you access the end iterator in the line above.

File: src/engine/world/world.h [hunk 4]
```diff
@@ -292,6 +292,15 @@ class World final
 
   void addPickupWidget(Sprite sprite, size_t count)
   {
+    auto isSameItemAndSameAction = [sprite](ui::PickupWidget item) -> bool {
```
Comment (stohrendorf) [inline src/engine/world/world.h:295]:
```suggestion
    auto isSameItemAndSameAction = [&sprite](const ui::PickupWidget& item) {
```
Avoid copies for non-primitive types unless necessary. Also, I think something like `isSameVirginItem` would better express what this predicate is doing.

**Review by stohrendorf — CHANGES_REQUESTED** [review]
All in all, a very good, straight-forward solution. I had a much more complicated one in mind.

File: src/ui/pickupwidget.h [hunk 5]
```diff
@@ -44,6 +44,27 @@ class PickupWidget
     return m_sprite.render1.x - m_sprite.render0.x;
   }
 
+  [[nodiscard]] auto getDuration() const
+  {
+    return m_duration;
+  }
+
+  [[nodiscard]] auto getSprite() const
```
Comment (stohrendorf) [inline src/ui/pickupwidget.h:52]:
```suggestion
  [[nodiscard]] const auto& getSprite() const
```

**Review by stohrendorf — CHANGES_REQUESTED** [review]
One last thing, then squash+rebase and merge.

By the way, maybe you want to have a look at fixup commits - they are a nice feature to basically edit commits by adding fixup commits, which will be integrated into the to-be-edited commits when squashing. The usual workflow is something like that:
* `git add -p` will allow you to select individual hunks to include in a commit. This is useful to split a bunch of changes into separate commits (see below for why this can be useful).
* `git commit --fixup <sha of commit to patch>` will add a fixup commit prefixed with `fixup!` and the message of the commit you want to patch.
* Do as many fixups as you wish.
* `git fetch origin/<branch>` of the branch you want to merge your changes into.
* `git rebase -i --autosquash origin/<branch>` will squash the `fixup!` commits into the to-be-patched commits. In the editor window that shows up then running the command, you'll notice `git` reordered the commits so the `fixup!` commits will be applied immediately after their corresponding commits and they will be squashed into them.
* `git push --force-with-lease` to force-push your commits.
* Merge :)

I hope this helps. Sounds more complicated, but if you have a series of commits in a larger PR, it makes life much easier. Unfortunately github doesn't seem to preserve the full history in the PR overview when using this workflow, because I used this recently in one of the multi-commit PRs and can't give you some example.

**Comment by serishema** [comment]
Fixup commits sound highly useful, I didn't know it was possible to in effect edit commits.
I've printed your comment about them and pinned it to the board behind my PC to remind me to use them. 

It should make it easier for you too if I can keep my branches mergeable by using them.

**Comment by stohrendorf** [comment]
Within "the git universe", it's usual to have one commit for one intention or one change. It's not a necessity, but especially within larger changes, it helps to associate changes with their corresponding commits. You're not required to make use of it, but it's certainly helpful after a review. Please keep in mind it's always helpful to regularily rebase on the target branch _without_ squashing to keep track of the changes, otherwise it's hard to follow the _actual_ changes and mark complaints as completed. (After all, change requests are complaints, no?) But I guess keeping track and rebasing within CE isn't that necessary as the changes are quite few. I've been working with a company where "stable master branch" was a necessity, but it's not a strong requirement in this case, so I wouldn't bother too much with being lax.
